daml-lf/spec/daml-lf-1.rst

```diff
@@ -3151,68 +3134,144 @@ in which keys were inserted into the map (insertion-order).
 * ``GENMAP_INSERT : ∀ α. ∀ β.  α → β → 'GenMap' α β → 'GenMap' α β``
 
   Inserts a new key and value in the map. If the key is already
-  present in the map, the associated value is replaced with the
-  supplied value, otherwise the new key/value entry is appended at the
-  ends of the map.
-
-  This raises an error if the key is not a valid map key. Keys are
-  compared according to ``~ᵥ``.
+  present according the builtin function ``EQUAL``, the associated
+  value is replaced with the supplied value, otherwise the key/value
+  is inserted in order according the builtin function ``LESS`` applied
+  on keys. This raises a runtime error if it tries to compare
+  incomparable values.
 
   [*Available in versions >= 1.dev*]
 
-* ``GENMAP_LOOKUP : ∀ α. ∀ β.  α → 'GenMap' α β → 'Optional' α``
+  Formally the builtin function ``GENMAP_INSERT`` semantics is defined
+  by the following rules. ::
 
-  Looks up the value at a key in the map.
+      𝕆('EQUAL' @σ vᵢ v) = Err t    for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertReplaceErr
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛 v w) = Err t
+
+      𝕆('EQUAL' @σ vᵢ v) = Ok 'True'    for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertReplace
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; …; vₙ ↦ wₙ〛 v w) =
+        'Ok' 〚v₁ ↦ w₁; …; vᵢ₋₁ ↦ wᵢ₋₁; vᵢ ↦ w;  vᵢ₊₁ ↦ wᵢ₊₁; …; vₙ ↦ wₙ〛
 
-  This raises an error if the key is not a valid map key. Keys are
-  compared according to the rules listed below.
+      𝕆('LESS' @σ vᵢ₋₁ v) = Err t    for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertInsertErrLeft
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; …; vₙ ↦ wₙ〛 v w) = Err t
+
+      𝕆('LESS' @σ v vᵢ) = Err t    for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertInsertErrRight
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; …; vₙ ↦ wₙ〛 v w) = Err t
+
+      𝕆('LESS' @σ v v₁) = Ok 'True'
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertInsertFirst
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; …; vₙ ↦ wₙ〛 v w) =
+        'Ok' 〚v ↦ w; v₁ ↦ w₁; …; vₙ ↦ wₙ〛
+
+      𝕆('LESS' @σ vᵢ₋₁ v) = Ok 'True'
+      𝕆('LESS' @σ v vᵢ) = Ok 'True'
+      for some i ∈ 2, …, n-1
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertInsertMiddle
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛 v w) =
+        'Ok' 〚v₁ ↦ w₁; … ; vᵢ₋₁ ↦ wᵢ₋₁; v ↦ w;  vᵢ ↦ wᵢ; … ; vₙ ↦ wₙ〛
+
+      𝕆('LESS' @σ vₙ v) = Ok 'True'
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertInsertLast
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; …; vₙ ↦ wₙ〛 v w) =
+        'Ok' 〚v₁ ↦ w₁; …; vₙ ↦ wₙ; v ↦ w〛
+
```

Comment:
We are missing a rule for insertion in an empty maps.

Again, for the empty map, I think the rule should perform a comparison in order to verify that the key is a comparable value.

Reply:
Good catch for the empty maps case